File: frontend/src/core/validation.ts
/**
 * Tip Güvenliği ve Doğrulama Sistemi
 * Tüm doğrulamalar tek yerden yönetilir
 */

import { logger } from './logger';
// Local constants for validation
const VALIDATION_CONSTANTS = {
  REGEX_PATTERNS: {
    PHONE: /^[\+]?[1-9][\d]{0,15}$/,
    URL: /^https?:\/\/.+/,
    PRODUCT_CODE: /^[A-Z0-9-]+$/,
    COLOR_CODE: /^[A-Z0-9]+$/
  },
  VALIDATION_RULES: {
    PASSWORD: {
      MIN_LENGTH: 8,
      MAX_LENGTH: 128
    },
    USERNAME: {
      MIN_LENGTH: 3,
      MAX_LENGTH: 50
    },
    BRAND_NAME: {
      MIN_LENGTH: 2,
      MAX_LENGTH: 100
    },
    PRODUCT_CODE: {
      MIN_LENGTH: 3,
      MAX_LENGTH: 20
    },
    TEMPLATE_NAME: {
      MIN_LENGTH: 3,
      MAX_LENGTH: 100
    },
    FILE: {
      MAX_SIZE_MB: 10
    }
  }
};

export enum ValidationType {
  REQUIRED = 'required',
  EMAIL = 'email',
  PHONE = 'phone',
  URL = 'url',
  PASSWORD = 'password',
  USERNAME = 'username',
  PRODUCT_CODE = 'product_code',
  COLOR_CODE = 'color_code',
  DATE = 'date',
  NUMBER = 'number',
  STRING = 'string',
  BOOLEAN = 'boolean',
  ARRAY = 'array',
  OBJECT = 'object',
  FILE = 'file',
  CUSTOM = 'custom',
}

export interface ValidationRule {
  field: string;
  validationType: ValidationType;
  required?: boolean;
  minLength?: number;
  maxLength?: number;
  minValue?: number;
  maxValue?: number;
  pattern?: RegExp;
  allowedValues?: any[];
  customValidator?: (value: any) => boolean;
  errorMessage?: string;
}

export interface ValidationResult {
  isValid: boolean;
  errors: string[];
  fieldErrors: Record<string, string[]>;
}

class Validator {
  private rules: Map<string, ValidationRule[]> = new Map();
  private errors: string[] = [];
  private fieldErrors: Map<string, string[]> = new Map();

  addRule(rule: ValidationRule): void {
    if (!this.rules.has(rule.field)) {
      this.rules.set(rule.field, []);
    }
    this.rules.get(rule.field)!.push(rule);
  }

  validate(data: Record<string, any>): ValidationResult {
    this.errors = [];
    this.fieldErrors.clear();

    for (const [field, rules] of Array.from(this.rules.entries())) {
      const value = data[field];
      const fieldErrors: string[] = [];

      for (const rule of rules) {
        if (!this.validateField(field, value, rule)) {
          const errorMessage = rule.errorMessage || this.getDefaultErrorMessage(field, rule);
          fieldErrors.push(errorMessage);
          this.errors.push(errorMessage);
        }
      }

      if (fieldErrors.length > 0) {
        this.fieldErrors.set(field, fieldErrors);
      }
    }

    return {
      isValid: this.errors.length === 0,
      errors: [...this.errors],
      fieldErrors: Object.fromEntries(this.fieldErrors),
    };
  }

  private validateField(field: string, value: any, rule: ValidationRule): boolean {
    // Required check
    if (rule.required && (value === null || value === undefined || value === '')) {
      return false;
    }

    // Skip validation if value is null/undefined and not required
    if ((value === null || value === undefined) && !rule.required) {
      return true;
    }

    // Type-specific validation
    switch (rule.validationType) {
      case ValidationType.EMAIL:
        return this.validateEmail(field, value, rule);
      case ValidationType.PHONE:
        return this.validatePhone(field, value, rule);
      case ValidationType.URL:
        return this.validateUrl(field, value, rule);
      case ValidationType.PASSWORD:
        return this.validatePassword(field, value, rule);
      case ValidationType.USERNAME:
        return this.validateUsername(field, value, rule);
      case ValidationType.PRODUCT_CODE:
        return this.validateProductCode(field, value, rule);
      case ValidationType.COLOR_CODE:
        return this.validateColorCode(field, value, rule);
      case ValidationType.DATE:
        return this.validateDate(field, value, rule);
      case ValidationType.NUMBER:
        return this.validateNumber(field, value, rule);
      case ValidationType.STRING:
        return this.validateString(field, value, rule);
      case ValidationType.BOOLEAN:
        return this.validateBoolean(field, value, rule);
      case ValidationType.ARRAY:
        return this.validateArray(field, value, rule);
      case ValidationType.OBJECT:
        return this.validateObject(field, value, rule);
      case ValidationType.FILE:
        return this.validateFile(field, value, rule);
      case ValidationType.CUSTOM:
        return this.validateCustom(field, value, rule);
      default:
        return true;
    }
  }

  private validateEmail(field: string, value: any, rule: ValidationRule): boolean {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return emailRegex.test(String(value));
  }

  private validatePhone(field: string, value: any, rule: ValidationRule): boolean {
    return VALIDATION_CONSTANTS.REGEX_PATTERNS.PHONE.test(String(value));
  }

  private validateUrl(field: string, value: any, rule: ValidationRule): boolean {
    return VALIDATION_CONSTANTS.REGEX_PATTERNS.URL.test(String(value));
  }

  private validatePassword(field: string, value: any, rule: ValidationRule): boolean {
    const password = String(value);

    if (password.length < VALIDATION_CONSTANTS.VALIDATION_RULES.PASSWORD.MIN_LENGTH) {
      return false;
    }

    if (password.length > VALIDATION_CONSTANTS.VALIDATION_RULES.PASSWORD.MAX_LENGTH) {
      return false;
    }

    // Check for at least one uppercase, lowercase, digit, and special character
    if (!/[A-Z]/.test(password)) return false;
    if (!/[a-z]/.test(password)) return false;
    if (!/\d/.test(password)) return false;
    if (!/[!@#$%^&*(),.?":{}|<>]/.test(password)) return false;

    return true;
  }

  private validateUsername(field: string, value: any, rule: ValidationRule): boolean {
    const username = String(value);

    if (username.length < VALIDATION_CONSTANTS.VALIDATION_RULES.USERNAME.MIN_LENGTH) {
      return false;
    }

    if (username.length > VALIDATION_CONSTANTS.VALIDATION_RULES.USERNAME.MAX_LENGTH) {
      return false;
    }

    // Check for valid characters (alphanumeric and underscore only)
    return /^[a-zA-Z0-9_]+$/.test(username);
  }

  private validateProductCode(field: string, value: any, rule: ValidationRule): boolean {
    return VALIDATION_CONSTANTS.REGEX_PATTERNS.PRODUCT_CODE.test(String(value));
  }

  private validateColorCode(field: string, value: any, rule: ValidationRule): boolean {
    return VALIDATION_CONSTANTS.REGEX_PATTERNS.COLOR_CODE.test(String(value));
  }

  private validateDate(field: string, value: any, rule: ValidationRule): boolean {
    try {
      const date = new Date(value);
      return !isNaN(date.getTime());
    } catch {
      return false;
    }
  }

  private validateNumber(field: string, value: any, rule: ValidationRule): boolean {
    const numValue = Number(value);

    if (isNaN(numValue)) {
      return false;
    }

    if (rule.minValue !== undefined && numValue < rule.minValue) {
      return false;
    }

    if (rule.maxValue !== undefined && numValue > rule.maxValue) {
      return false;
    }

    return true;
  }

  private validateString(field: string, value: any, rule: ValidationRule): boolean {
    const strValue = String(value);

    if (rule.minLength !== undefined && strValue.length < rule.minLength) {
      return false;
    }

    if (rule.maxLength !== undefined && strValue.length > rule.maxLength) {
      return false;
    }

    if (rule.pattern && !rule.pattern.test(strValue)) {
      return false;
    }

    if (rule.allowedValues && !rule.allowedValues.includes(strValue)) {
      return false;
    }

    return true;
  }

  private validateBoolean(field: string, value: any, rule: ValidationRule): boolean {
    return typeof value === 'boolean';
  }

  private validateArray(field: string, value: any, rule: ValidationRule): boolean {
    if (!Array.isArray(value)) {
      return false;
    }

    if (rule.minLength !== undefined && value.length < rule.minLength) {
      return false;
    }

    if (rule.maxLength !== undefined && value.length > rule.maxLength) {
      return false;
    }

    return true;
  }

  private validateObject(field: string, value: any, rule: ValidationRule): boolean {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
  }

  private validateFile(field: string, value: any, rule: ValidationRule): boolean {
    return value instanceof File;
  }

  private validateCustom(field: string, value: any, rule: ValidationRule): boolean {
    if (rule.customValidator) {
      try {
        return rule.customValidator(value);
      } catch (error) {
        logger.error('Custom validation error', { field, error });
        return false;
      }
    }
    return true;
  }

  private getDefaultErrorMessage(field: string, rule: ValidationRule): string {
    switch (rule.validationType) {
      case ValidationType.REQUIRED:
        return `${field} is required`;
      case ValidationType.EMAIL:
        return `${field} must be a valid email address`;
      case ValidationType.PHONE:
        return `${field} must be a valid phone number`;
      case ValidationType.URL:
        return `${field} must be a valid URL`;
      case ValidationType.PASSWORD:
        return `${field} must meet password requirements`;
      case ValidationType.USERNAME:
        return `${field} must be a valid username`;
      case ValidationType.PRODUCT_CODE:
        return `${field} must be a valid product code`;
      case ValidationType.COLOR_CODE:
        return `${field} must be a valid color code`;
      case ValidationType.DATE:
        return `${field} must be a valid date`;
      case ValidationType.NUMBER:
        return `${field} must be a valid number`;
      case ValidationType.STRING:
        return `${field} must be a valid string`;
      case ValidationType.BOOLEAN:
        return `${field} must be a boolean value`;
      case ValidationType.ARRAY:
        return `${field} must be an array`;
      case ValidationType.OBJECT:
        return `${field} must be an object`;
      case ValidationType.FILE:
        return `${field} must be a file`;
      default:
        return `${field} is invalid`;
    }
  }

  getErrors(): string[] {
    return [...this.errors];
  }

  getFieldErrors(): Record<string, string[]> {
    return Object.fromEntries(this.fieldErrors);
  }

  clearErrors(): void {
    this.errors = [];
    this.fieldErrors.clear();
  }
}

// Specific validators
class UserValidator extends Validator {
  constructor() {
    super();

    this.addRule({
      field: 'username',
      validationType: ValidationType.USERNAME,
      required: true,
      minLength: VALIDATION_CONSTANTS.VALIDATION_RULES.USERNAME.MIN_LENGTH,
      maxLength: VALIDATION_CONSTANTS.VALIDATION_RULES.USERNAME.MAX_LENGTH,
    });

    this.addRule({
      field: 'email',
      validationType: ValidationType.EMAIL,
      required: true,
    });

    this.addRule({
      field: 'password',
      validationType: ValidationType.PASSWORD,
      required: true,
      minLength: VALIDATION_CONSTANTS.VALIDATION_RULES.PASSWORD.MIN_LENGTH,
      maxLength: VALIDATION_CONSTANTS.VALIDATION_RULES.PASSWORD.MAX_LENGTH,
    });

    this.addRule({
      field: 'role',
      validationType: ValidationType.STRING,
      required: true,
      allowedValues: ['super_admin', 'brand_manager', 'employee', 'viewer'],
    });

    this.addRule({
      field: 'status',
      validationType: ValidationType.STRING,
      required: true,
      allowedValues: ['active', 'inactive', 'pending', 'suspended'],
    });
  }
}

class BrandValidator extends Validator {
  constructor() {
    super();

    this.addRule({
      field: 'name',
      validationType: ValidationType.STRING,
      required: true,
      minLength: VALIDATION_CONSTANTS.VALIDATION_RULES.BRAND_NAME.MIN_LENGTH,
      maxLength: VALIDATION_CONSTANTS.VALIDATION_RULES.BRAND_NAME.MAX_LENGTH,
    });

    this.addRule({
      field: 'description',
      validationType: ValidationType.STRING,
      required: false,
      maxLength: 500,
    });

    this.addRule({
      field: 'status',
      validationType: ValidationType.STRING,
      required: true,
      allowedValues: ['active', 'inactive', 'pending'],
    });
  }
}

class ProductValidator extends Validator {
  constructor() {
    super();

    this.addRule({
      field: 'name',
      validationType: ValidationType.STRING,
      required: true,
      minLength: 3,
      maxLength: 100,
    });

    this.addRule({
      field: 'code',
      validationType: ValidationType.PRODUCT_CODE,
      required: true,
      minLength: VALIDATION_CONSTANTS.VALIDATION_RULES.PRODUCT_CODE.MIN_LENGTH,
      maxLength: VALIDATION_CONSTANTS.VALIDATION_RULES.PRODUCT_CODE.MAX_LENGTH,
    });

    this.addRule({
      field: 'color',
      validationType: ValidationType.STRING,
      required: false,
      maxLength: 50,
    });

    this.addRule({
      field: 'brand_id',
      validationType: ValidationType.NUMBER,
      required: true,
      minValue: 1,
    });
  }
}

class TemplateValidator extends Validator {
  constructor() {
    super();

    this.addRule({
      field: 'name',
      validationType: ValidationType.STRING,
      required: true,
      minLength: VALIDATION_CONSTANTS.VALIDATION_RULES.TEMPLATE_NAME.MIN_LENGTH,
      maxLength: VALIDATION_CONSTANTS.VALIDATION_RULES.TEMPLATE_NAME.MAX_LENGTH,
    });

    this.addRule({
      field: 'description',
      validationType: ValidationType.STRING,
      required: false,
      maxLength: 500,
    });

    this.addRule({
      field: 'template_type',
      validationType: ValidationType.STRING,
      required: true,
      allowedValues: ['standard', 'modern', 'classic', 'custom'],
    });

    this.addRule({
      field: 'product_id',
      validationType: ValidationType.NUMBER,
      required: true,
      minValue: 1,
    });

    this.addRule({
      field: 'brand_id',
      validationType: ValidationType.NUMBER,
      required: true,
      minValue: 1,
    });
  }
}

class FileValidator extends Validator {
  constructor() {
    super();

    this.addRule({
      field: 'file',
      validationType: ValidationType.FILE,
      required: true,
    });

    this.addRule({
      field: 'maxSize',
      validationType: ValidationType.NUMBER,
      required: false,
      maxValue: VALIDATION_CONSTANTS.VALIDATION_RULES.FILE.MAX_SIZE_MB * 1024 * 1024, // Convert to bytes
    });

    this.addRule({
      field: 'allowedTypes',
      validationType: ValidationType.ARRAY,
      required: false,
    });
  }
}

// Global validators
export const userValidator = new UserValidator();
export const brandValidator = new BrandValidator();
export const productValidator = new ProductValidator();
export const templateValidator = new TemplateValidator();
export const fileValidator = new FileValidator();

// Validation functions
export function validateUser(data: Record<string, any>): ValidationResult {
  return userValidator.validate(data);
}

export function validateBrand(data: Record<string, any>): ValidationResult {
  return brandValidator.validate(data);
}

export function validateProduct(data: Record<string, any>): ValidationResult {
  return productValidator.validate(data);
}

export function validateTemplate(data: Record<string, any>): ValidationResult {
  return templateValidator.validate(data);
}

export function validateFile(data: Record<string, any>): ValidationResult {
  return fileValidator.validate(data);
}

// Utility functions
export function createValidator(rules: ValidationRule[]): Validator {
  const validator = new Validator();
  rules.forEach(rule => validator.addRule(rule));
  return validator;
}

export function validateFormData(formData: FormData, validator: Validator): ValidationResult {
  const data: Record<string, any> = {};
  
  for (const [key, value] of Array.from(formData.entries())) {
    data[key] = value;
  }
  
  return validator.validate(data);
}

export function validateObjectData(data: Record<string, any>, validator: Validator): ValidationResult {
  return validator.validate(data);
}

export function getValidationErrors(result: ValidationResult): string[] {
  return result.errors;
}

export function getFieldValidationErrors(result: ValidationResult, field: string): string[] {
  return result.fieldErrors[field] || [];
}

export function hasValidationErrors(result: ValidationResult): boolean {
  return !result.isValid;
}

export function hasFieldValidationErrors(result: ValidationResult, field: string): boolean {
  return result.fieldErrors[field] && result.fieldErrors[field].length > 0;
}

export default {
  Validator,
  UserValidator,
  BrandValidator,
  ProductValidator,
  TemplateValidator,
  FileValidator,
  validateUser,
  validateBrand,
  validateProduct,
  validateTemplate,
  validateFile,
  createValidator,
  validateFormData,
  validateObjectData,
  getValidationErrors,
  getFieldValidationErrors,
  hasValidationErrors,
  hasFieldValidationErrors,
};

File: backend/core/validation.py
"""
Tip Güvenliği ve Doğrulama Sistemi
Tüm doğrulamalar tek yerden yönetilir
"""

import re
import email_validator
from typing import Any, Dict, List, Optional, Union, Type, Callable
from dataclasses import dataclass
from enum import Enum
from pydantic import BaseModel, ValidationError as PydanticValidationError
from core.logging import get_logger
from core.exceptions import ValidationError
from core.constants import RegexPatterns, ValidationRules

logger = get_logger('validation')

class ValidationType(str, Enum):
    """Doğrulama türleri"""
    REQUIRED = "required"
    EMAIL = "email"
    PHONE = "phone"
    URL = "url"
    PASSWORD = "password"
    USERNAME = "username"
    PRODUCT_CODE = "product_code"
    COLOR_CODE = "color_code"
    DATE = "date"
    NUMBER = "number"
    STRING = "string"
    BOOLEAN = "boolean"
    ARRAY = "array"
    OBJECT = "object"
    FILE = "file"
    CUSTOM = "custom"

@dataclass
class ValidationRule:
    """Doğrulama kuralı"""
    field: str
    validation_type: ValidationType
    required: bool = False
    min_length: Optional[int] = None
    max_length: Optional[int] = None
    min_value: Optional[Union[int, float]] = None
    max_value: Optional[Union[int, float]] = None
    pattern: Optional[str] = None
    allowed_values: Optional[List[Any]] = None
    custom_validator: Optional[Callable] = None
    error_message: Optional[str] = None

class Validator:
    """Ana doğrulayıcı sınıf"""
    
    def __init__(self):
        self.rules: Dict[str, List[ValidationRule]] = {}
        self.errors: List[str] = []
    
    def add_rule(self, rule: ValidationRule):
        """Doğrulama kuralı ekle"""
        if rule.field not in self.rules:
            self.rules[rule.field] = []
        self.rules[rule.field].append(rule)
    
    def validate(self, data: Dict[str, Any]) -> bool:
        """Veriyi doğrula"""
        self.errors = []
        
        for field, rules in self.rules.items():
            value = data.get(field)
            
            for rule in rules:
                if not self._validate_field(field, value, rule):
                    return False
        
        return len(self.errors) == 0
    
    def _validate_field(self, field: str, value: Any, rule: ValidationRule) -> bool:
        """Alan doğrulaması"""
        # Required check
        if rule.required and (value is None or value == ""):
            error_msg = rule.error_message or f"{field} is required"
            self.errors.append(error_msg)
            return False
        
        # Skip validation if value is None and not required
        if value is None and not rule.required:
            return True
        
        # Type-specific validation
        if rule.validation_type == ValidationType.EMAIL:
            return self._validate_email(field, value, rule)
        elif rule.validation_type == ValidationType.PHONE:
            return self._validate_phone(field, value, rule)
        elif rule.validation_type == ValidationType.URL:
            return self._validate_url(field, value, rule)
        elif rule.validation_type == ValidationType.PASSWORD:
            return self._validate_password(field, value, rule)
        elif rule.validation_type == ValidationType.USERNAME:
            return self._validate_username(field, value, rule)
        elif rule.validation_type == ValidationType.PRODUCT_CODE:
            return self._validate_product_code(field, value, rule)
        elif rule.validation_type == ValidationType.COLOR_CODE:
            return self._validate_color_code(field, value, rule)
        elif rule.validation_type == ValidationType.DATE:
            return self._validate_date(field, value, rule)
        elif rule.validation_type == ValidationType.NUMBER:
            return self._validate_number(field, value, rule)
        elif rule.validation_type == ValidationType.STRING:
            return self._validate_string(field, value, rule)
        elif rule.validation_type == ValidationType.BOOLEAN:
            return self._validate_boolean(field, value, rule)
        elif rule.validation_type == ValidationType.ARRAY:
            return self._validate_array(field, value, rule)
        elif rule.validation_type == ValidationType.OBJECT:
            return self._validate_object(field, value, rule)
        elif rule.validation_type == ValidationType.FILE:
            return self._validate_file(field, value, rule)
        elif rule.validation_type == ValidationType.CUSTOM:
            return self._validate_custom(field, value, rule)
        
        return True
    
    def _validate_email(self, field: str, value: Any, rule: ValidationRule) -> bool:
        """Email doğrulaması"""
        try:
            email_validator.validate_email(value)
            return True
        except email_validator.EmailNotValidError:
            error_msg = rule.error_message or f"{field} is not a valid email"
            self.errors.append(error_msg)
            return False
    
    def _validate_phone(self, field: str, value: Any, rule: ValidationRule) -> bool:
        """Telefon doğrulaması"""
        if not re.match(RegexPatterns.PHONE, str(value)):
            error_msg = rule.error_message or f"{field} is not a valid phone number"
            self.errors.append(error_msg)
            return False
        return True
    
    def _validate_url(self, field: str, value: Any, rule: ValidationRule) -> bool:
        """URL doğrulaması"""
        if not re.match(RegexPatterns.URL, str(value)):
            error_msg = rule.error_message or f"{field} is not a valid URL"
            self.errors.append(error_msg)
            return False
        return True
    
    def _validate_password(self, field: str, value: Any, rule: ValidationRule) -> bool:
        """Şifre doğrulaması"""
        password = str(value)
        
        if len(password) < ValidationRules.PASSWORD_MIN_LENGTH:
            error_msg = rule.error_message or f"{field} must be at least {ValidationRules.PASSWORD_MIN_LENGTH} characters"
            self.errors.append(error_msg)
            return False
        
        if len(password) > ValidationRules.PASSWORD_MAX_LENGTH:
            error_msg = rule.error_message or f"{field} must be at most {ValidationRules.PASSWORD_MAX_LENGTH} characters"
            self.errors.append(error_msg)
            return False
        
        # Check for at least one uppercase, lowercase, digit, and special character
        if not re.search(r'[A-Z]', password):
            error_msg = rule.error_message or f"{field} must contain at least one uppercase letter"
            self.errors.append(error_msg)
            return False
        
        if not re.search(r'[a-z]', password):
            error_msg = rule.error_message or f"{field} must contain at least one lowercase letter"
            self.errors.append(error_msg)
            return False
        
        if not re.search(r'\d', password):
            error_msg = rule.error_message or f"{field} must contain at least one digit"
            self.errors.append(error_msg)
            return False
        
        if not re.search(r'[!@#$%^&*(),.?":{}|<>]', password):
            error_msg = rule.error_message or f"{field} must contain at least one special character"
            self.errors.append(error_msg)
            return False
        
        return True
    
    def _validate_username(self, field: str, value: Any, rule: ValidationRule) -> bool:
        """Kullanıcı adı doğrulaması"""
        username = str(value)
        
        if len(username) < ValidationRules.USERNAME_MIN_LENGTH:
            error_msg = rule.error_message or f"{field} must be at least {ValidationRules.USERNAME_MIN_LENGTH} characters"
            self.errors.append(error_msg)
            return False
        
        if len(username) > ValidationRules.USERNAME_MAX_LENGTH:
            error_msg = rule.error_message or f"{field} must be at most {ValidationRules.USERNAME_MAX_LENGTH} characters"
            self.errors.append(error_msg)
            return False
        
        # Check for valid characters (alphanumeric and underscore only)
        if not re.match(r'^[a-zA-Z0-9_]+$', username):
            error_msg = rule.error_message or f"{field} can only contain letters, numbers, and underscores"
            self.errors.append(error_msg)
            return False
        
        return True
    
    def _validate_product_code(self, field: str, value: Any, rule: ValidationRule) -> bool:
        """Ürün kodu doğrulaması"""
        if not re.match(RegexPatterns.PRODUCT_CODE, str(value)):
            error_msg = rule.error_message or f"{field} is not a valid product code"
            self.errors.append(error_msg)
            return False
        return True
    
    def _validate_color_code(self, field: str, value: Any, rule: ValidationRule) -> bool:
        """Renk kodu doğrulaması"""
        if not re.match(RegexPatterns.COLOR_CODE, str(value)):
            error_msg = rule.error_message or f"{field} is not a valid color code"
            self.errors.append(error_msg)
            return False
        return True
    
    def _validate_date(self, field: str, value: Any, rule: ValidationRule) -> bool:
        """Tarih doğrulaması"""
        try:
            from datetime import datetime
            if isinstance(value, str):
                datetime.fromisoformat(value)
            elif isinstance(value, datetime):
                pass
            else:
                raise ValueError("Invalid date format")
            return True
        except (ValueError, TypeError):
            error_msg = rule.error_message or f"{field} is not a valid date"
            self.errors.append(error_msg)
            return False
    
    def _validate_number(self, field: str, value: Any, rule: ValidationRule) -> bool:
        """Sayı doğrulaması"""
        try:
            num_value = float(value)
            
            if rule.min_value is not None and num_value < rule.min_value:
                error_msg = rule.error_message or f"{field} must be at least {rule.min_value}"
                self.errors.append(error_msg)
                return False
            
            if rule.max_value is not None and num_value > rule.max_value:
                error_msg = rule.error_message or f"{field} must be at most {rule.max_value}"
                self.errors.append(error_msg)
                return False
            
            return True
        except (ValueError, TypeError):
            error_msg = rule.error_message or f"{field} is not a valid number"
            self.errors.append(error_msg)
            return False
    
    def _validate_string(self, field: str, value: Any, rule: ValidationRule) -> bool:
        """String doğrulaması"""
        str_value = str(value)
        
        if rule.min_length is not None and len(str_value) < rule.min_length:
            error_msg = rule.error_message or f"{field} must be at least {rule.min_length} characters"
            self.errors.append(error_msg)
            return False
        
        if rule.max_length is not None and len(str_value) > rule.max_length:
            error_msg = rule.error_message or f"{field} must be at most {rule.max_length} characters"
            self.errors.append(error_msg)
            return False
        
        if rule.pattern and not re.match(rule.pattern, str_value):
            error_msg = rule.error_message or f"{field} does not match required pattern"
            self.errors.append(error_msg)
            return False
        
        if rule.allowed_values and str_value not in rule.allowed_values:
            error_msg = rule.error_message or f"{field} must be one of {rule.allowed_values}"
            self.errors.append(error_msg)
            return False
        
        return True
    
    def _validate_boolean(self, field: str, value: Any, rule: ValidationRule) -> bool:
        """Boolean doğrulaması"""
        if not isinstance(value, bool):
            error_msg = rule.error_message or f"{field} must be a boolean value"
            self.errors.append(error_msg)
            return False
        return True
    
    def _validate_array(self, field: str, value: Any, rule: ValidationRule) -> bool:
        """Array doğrulaması"""
        if not isinstance(value, list):
            error_msg = rule.error_message or f"{field} must be an array"
            self.errors.append(error_msg)
            return False
        
        if rule.min_length is not None and len(value) < rule.min_length:
            error_msg = rule.error_message or f"{field} must have at least {rule.min_length} items"
            self.errors.append(error_msg)
            return False
        
        if rule.max_length is not None and len(value) > rule.max_length:
            error_msg = rule.error_message or f"{field} must have at most {rule.max_length} items"
            self.errors.append(error_msg)
            return False
        
        return True
    
    def _validate_object(self, field: str, value: Any, rule: ValidationRule) -> bool:
        """Object doğrulaması"""
        if not isinstance(value, dict):
            error_msg = rule.error_message or f"{field} must be an object"
            self.errors.append(error_msg)
            return False
        return True
    
    def _validate_file(self, field: str, value: Any, rule: ValidationRule) -> bool:
        """Dosya doğrulaması"""
        # This would typically check file size, type, etc.
        # For now, just check if it's a file-like object
        if not hasattr(value, 'read'):
            error_msg = rule.error_message or f"{field} must be a file"
            self.errors.append(error_msg)
            return False
        return True
    
    def _validate_custom(self, field: str, value: Any, rule: ValidationRule) -> bool:
        """Özel doğrulama"""
        if rule.custom_validator:
            try:
                return rule.custom_validator(value)
            except Exception as e:
                error_msg = rule.error_message or f"{field} validation failed: {str(e)}"
                self.errors.append(error_msg)
                return False
        return True
    
    def get_errors(self) -> List[str]:
        """Hataları al"""
        return self.errors.copy()
    
    def clear_errors(self):
        """Hataları temizle"""
        self.errors = []

# Pydantic model validators
class BaseValidatorModel(BaseModel):
    """Temel doğrulama modeli"""
    
    class Config:
        validate_assignment = True
        use_enum_values = True
        extra = "forbid"
    
    def validate_field(self, field_name: str, value: Any) -> bool:
        """Alan doğrulaması"""
        try:
            setattr(self, field_name, value)
            return True
        except PydanticValidationError as e:
            logger.error(f"Validation error for field {field_name}: {e}")
            return False

# Specific validators
class UserValidator(BaseValidatorModel):
    """Kullanıcı doğrulayıcı"""
    username: str
    email: str
    password: str
    role: str
    status: str
    
    @classmethod
    def create_validator(cls) -> Validator:
        """Kullanıcı doğrulayıcı oluştur"""
        validator = Validator()
        
        validator.add_rule(ValidationRule(
            field="username",
            validation_type=ValidationType.USERNAME,
            required=True,
            min_length=ValidationRules.USERNAME_MIN_LENGTH,
            max_length=ValidationRules.USERNAME_MAX_LENGTH
        ))
        
        validator.add_rule(ValidationRule(
            field="email",
            validation_type=ValidationType.EMAIL,
            required=True
        ))
        
        validator.add_rule(ValidationRule(
            field="password",
            validation_type=ValidationType.PASSWORD,
            required=True,
            min_length=ValidationRules.PASSWORD_MIN_LENGTH,
            max_length=ValidationRules.PASSWORD_MAX_LENGTH
        ))
        
        validator.add_rule(ValidationRule(
            field="role",
            validation_type=ValidationType.STRING,
            required=True,
            allowed_values=["super_admin", "brand_manager", "employee", "viewer"]
        ))
        
        validator.add_rule(ValidationRule(
            field="status",
            validation_type=ValidationType.STRING,
            required=True,
            allowed_values=["active", "inactive", "pending", "suspended"]
        ))
        
        return validator

class BrandValidator(BaseValidatorModel):
    """Marka doğrulayıcı"""
    name: str
    description: Optional[str] = None
    status: str
    
    @classmethod
    def create_validator(cls) -> Validator:
        """Marka doğrulayıcı oluştur"""
        validator = Validator()
        
        validator.add_rule(ValidationRule(
            field="name",
            validation_type=ValidationType.STRING,
            required=True,
            min_length=ValidationRules.BRAND_NAME_MIN_LENGTH,
            max_length=ValidationRules.BRAND_NAME_MAX_LENGTH
        ))
        
        validator.add_rule(ValidationRule(
            field="description",
            validation_type=ValidationType.STRING,
            required=False,
            max_length=500
        ))
        
        validator.add_rule(ValidationRule(
            field="status",
            validation_type=ValidationType.STRING,
            required=True,
            allowed_values=["active", "inactive", "pending"]
        ))
        
        return validator

class ProductValidator(BaseValidatorModel):
    """Ürün doğrulayıcı"""
    name: str
    code: str
    color: Optional[str] = None
    brand_id: int
    
    @classmethod
    def create_validator(cls) -> Validator:
        """Ürün doğrulayıcı oluştur"""
        validator = Validator()
        
        validator.add_rule(ValidationRule(
            field="name",
            validation_type=ValidationType.STRING,
            required=True,
            min_length=3,
            max_length=100
        ))
        
        validator.add_rule(ValidationRule(
            field="code",
            validation_type=ValidationType.PRODUCT_CODE,
            required=True,
            min_length=ValidationRules.PRODUCT_CODE_MIN_LENGTH,
            max_length=ValidationRules.PRODUCT_CODE_MAX_LENGTH
        ))
        
        validator.add_rule(ValidationRule(
            field="color",
            validation_type=ValidationType.STRING,
            required=False,
            max_length=50
        ))
        
        validator.add_rule(ValidationRule(
            field="brand_id",
            validation_type=ValidationType.NUMBER,
            required=True,
            min_value=1
        ))
        
        return validator

# Global validators
user_validator = UserValidator.create_validator()
brand_validator = BrandValidator.create_validator()
product_validator = ProductValidator.create_validator()

def validate_user(data: Dict[str, Any]) -> bool:
    """Kullanıcı doğrulaması"""
    return user_validator.validate(data)

def validate_brand(data: Dict[str, Any]) -> bool:
    """Marka doğrulaması"""
    return brand_validator.validate(data)

def validate_product(data: Dict[str, Any]) -> bool:
    """Ürün doğrulaması"""
    return product_validator.validate(data)

def get_validation_errors() -> List[str]:
    """Doğrulama hatalarını al"""
    all_errors = []
    all_errors.extend(user_validator.get_errors())
    all_errors.extend(brand_validator.get_errors())
    all_errors.extend(product_validator.get_errors())
    return all_errors

def clear_validation_errors():
    """Doğrulama hatalarını temizle"""
    user_validator.clear_errors()
    brand_validator.clear_errors()
    product_validator.clear_errors()
